lib/error_report_notify.c

```diff
@@ -0,0 +1,207 @@
+}
+static DECLARE_DELAYED_WORK(reporting_done, error_report_notify);
+
+static int get_encoded_pid(void)
+{
+	return (current->pid << 1) | (!!in_task());
+}
+
+/*
+ * Trace hook for the error_report_start event. In an unlikely case of another
+ * task already printing a report bail out, otherwise save the current pid
+ * together with in_task() return value.
+ */
+static void probe_report_start(void *ignore, const char *tool_name,
+			       unsigned long id)
+{
+	/* Acquire the writer lock. */
+	if (atomic_cmpxchg_relaxed(&current_pid, -1, get_encoded_pid()) != -1)
+		return;
+	current_report_pos = 0;
+}
+
+/*
+ * Trace hook for the error_report_end event. If an event from the mismatching
+ * error_report_start is received, it is ignored. Otherwise, null-terminate the
+ * buffer, increase the report count (effectively releasing the report to
+ * last_report_show() and schedule a notification about a new report.
+ */
+static void probe_report_end(void *ignore, const char *tool_name,
+			     unsigned long id)
+{
+	pid_t pid = atomic_read(&current_pid);
+	int idx;
+
+	if (pid != get_encoded_pid())
+		return;
+
+	idx = (atomic_read_acquire(&num_reports) + 1) % 2;
+	if (current_report_pos == BUF_SIZE)
+		report_buffer[idx][current_report_pos - 1] = 0;
+	else
+		report_buffer[idx][current_report_pos] = 0;
+
+	atomic_inc_return_release(&num_reports);
+	schedule_delayed_work(&reporting_done, 0);
+	/* Release the writer lock */
+	atomic_set(&current_pid, -1);
+}
+
+/*
+ * Trace hook for the console event. If a line comes from a task/CPU that did
+ * not send the error_report_start event, that line is ignored. Otherwise, it
+ * is stored in the report_buffer[(num_reports + 1) % 2].
+ */
+static void probe_console(void *ignore, const char *buf, size_t len)
+{
+	int pid = atomic_read(&current_pid);
+	size_t to_copy;
+	int idx;
+
+	if (pid != get_encoded_pid() || truncated)
+		return;
+
+	idx = (atomic_read_acquire(&num_reports) + 1) % 2;
+	to_copy = min(len, BUF_SIZE - current_report_pos);
+	memcpy(report_buffer[idx] + current_report_pos, buf, to_copy);
+	current_report_pos += to_copy;
+	if (len > to_copy) {
+		truncated = true;
+		memcpy(report_buffer[idx] + current_report_pos -
+			       sizeof(TRUNC_MSG),
+		       TRUNC_MSG, sizeof(TRUNC_MSG));
+	}
+}
+
+static void register_tracepoints(void)
+{
+	register_trace_console(probe_console, NULL);
+	register_trace_error_report_start(probe_report_start, NULL);
+	register_trace_error_report_end(probe_report_end, NULL);
+}
+
+/*
+ * read() handler for /sys/kernel/error_report/last_report.
+ * Because the number of reports can change under our feet, check it again
+ * after copying the report, and retry if the numbers mismatch.
+ * */
+static ssize_t last_report_show(struct kobject *kobj,
+				struct kobj_attribute *attr, char *buf)
+{
+	int index = atomic_read_acquire(&num_reports), old_index;
+	char *out_buf = vmalloc(BUF_SIZE);
+	ssize_t ret;
+
+	do {
+		old_index = index;
```

Comment:
Updated the comment above: we were missing an "smp_rmb()" because the memcpy itself must be ordered before the re-check.

Normal seqlocks also add this. And C++11 seqlocks would use a seq_cst read of some sort. But we don't have a fully-ordered op, and doing a 0-increment would feel wrong.